### DISCUSSION TITLE
How to use kyuubi(kyuubi-1.2.0-bin-spark-3.1-hadoop3.2) with delta-lake (Azure blob storage)

### DISCUSSION THREAD

**Discussion by zhichaoleo** [apache/kyuubi#1453] [Q&A] [answered]
Hi Team.
We are trying to use kyuubi(kyuubi-1.2.0-bin-spark-3.1-hadoop3.2) with delta-lake (Azure blob storage). But I can't find any documents to set up related configurations on kyuubi offical doc site or google search.
My kyuubi-defaults.conf:

spark.sql.extensions=io.delta.sql.DeltaSparkSessionExtension
spark.sql.catalog.spark_catalog=org.apache.spark.sql.delta.catalog.DeltaCatalog
fs.azure.account.key.<account>.blob.core.windows.net=<token>

But it doesn't work as expected. Am I configure kyuubi right? Is there any docs and examples about how to use kyuubi with delta-lake (Azure blob storage) ?
Thanks!

**Comment by pan3793** [comment, marked as answer]
1.2.0 is not the latest version, we move the release to the https://kyuubi.apache.org/releases.html after entering apache incubator.

For delta integration, please follow the delta official doc https://docs.delta.io/latest/quick-start.html to setup detla jar and spark confs, the basic setup steps should be:
1. Download delta jar, and put the jar into `$SPARK_HOME/jars/`
2. Modify `$SPARK_HOME/conf/spark-defaults.conf` and put the following items.
```
spark.sql.extensions o.delta.sql.DeltaSparkSessionExtension
spark.sql.catalog.spark_catalog org.apache.spark.sql.delta.catalog.DeltaCatalog
# azure confs ...
```
3. Use `spark-shell` to verify it works.
4. Deploy Kyuubi
5. Use client (e.g. beeline) to connect Kyuubi and run SQL to opreate delta tables.

For deployment, we recommand persist the required confs into `$SPARK_HOME/conf/spark-defaults.conf`, which will take effect for `spark-submit`, `spark-shell` and Kyuubi, and overwrite those confs specfic to Kyuubi in `$KYUUBI_HOME/conf/kyuubi-defaults.conf`.

**Reply by zhichaoleo** [reply]
Hi @pan3793 .
Thanks for your help! I change my kyuubi to the latest version _apache-kyuubi-1.3.1-incubating-bin_. 
And my spark can connect to Azure delta-lake successfully:
```
hadoop@pvgl56604422a:/usr/apache/current/spark/bin> spark-shell
21/11/25 17:01:51 WARN Utils: Your hostname, ******** resolves to a loopback address: 127.0.0.1; using ***** instead (on interface eth0)
21/11/25 17:01:51 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
21/11/25 17:01:52 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context Web UI available at http://********:4040
Spark context available as 'sc' (master = spark://***********:7077, app id = app-20211125170157-0001).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.1.2
      /_/

Using Scala version 2.12.10 (OpenJDK 64-Bit Server VM, Java 1.8.0_302)
Type in expressions to have them evaluated.
Type :help for more information.

scala> val df01=spark.read.format("delta").load("wasbs://demo@********.blob.core.windows.net/alexDemo20211125")
21/11/25 17:02:35 WARN MetricsConfig: Cannot locate configuration: tried hadoop-metrics2-azure-file-system.properties,hadoop-metrics2.properties
21/11/25 17:02:59 WARN TaskSchedulerImpl: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
21/11/25 17:03:14 WARN TaskSchedulerImpl: Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources
df01: org.apache.spark.sql.DataFrame = [id: bigint]

scala> df01.show()
+----+
|  id|
+----+
|1000|
|1001|
|1002|
|1003|
|1004|
|1005|
|1006|
|1007|
|1008|
|1009|
|1010|
|1011|
|1012|
|1013|
|1014|
|1015|
|1016|
|1017|
|1018|
|1019|
+----+
only showing top 20 rows
```


But I got an error at step 5, I start up kyuubi and  use beeline to connect kyuubi, and the error log is :
```
Starting Kyuubi Server from /usr/apache/current/kyuubi
Warn: Not find kyuubi environment file /usr/apache/current/kyuubi/conf/kyuubi-env.sh, using default ones...
JAVA_HOME: /usr/lib64/jvm/java
KYUUBI_HOME: /usr/apache/current/kyuubi
KYUUBI_CONF_DIR: /usr/apache/current/kyuubi/conf
KYUUBI_LOG_DIR: /usr/apache/current/kyuubi/logs
KYUUBI_PID_DIR: /usr/apache/current/kyuubi/pid
KYUUBI_WORK_DIR_ROOT: /usr/apache/current/kyuubi/work
SPARK_HOME: /usr/apache/current/spark
SPARK_CONF_DIR: /usr/apache/current/spark/conf
HADOOP_CONF_DIR:
Starting org.apache.kyuubi.server.KyuubiServer, logging to /usr/apache/current/kyuubi/logs/kyuubi-hadoop-org.apache.kyuubi.server.KyuubiServer-pvgl56604422a.out
Welcome to
  __  __                           __
 /\ \/\ \                         /\ \      __
 \ \ \/'/'  __  __  __  __  __  __\ \ \____/\_\
  \ \ , <  /\ \/\ \/\ \/\ \/\ \/\ \\ \ '__`\/\ \
   \ \ \\`\\ \ \_\ \ \ \_\ \ \ \_\ \\ \ \L\ \ \ \
    \ \_\ \_\/`____ \ \____/\ \____/ \ \_,__/\ \_\
     \/_/\/_/`/___/> \/___/  \/___/   \/___/  \/_/
                /\___/
                \/__/
```

```
2021-11-25 16:38:37.602 WARN service.ThriftFrontendService: Error opening session:
org.apache.kyuubi.KyuubiSQLException: Error opening session SessionHandle [488765f3-24da-4622-8a1c-0641574d1c10] for anonymous due to java.net.ConnectException: Connection refused (Connection refused)
        at org.apache.kyuubi.KyuubiSQLException$.apply(KyuubiSQLException.scala:68)
        at org.apache.kyuubi.session.KyuubiSessionManager.openSession(KyuubiSessionManager.scala:74)
        at org.apache.kyuubi.service.AbstractBackendService.openSession(AbstractBackendService.scala:45)
        at org.apache.kyuubi.service.ThriftFrontendService.getSessionHandle(ThriftFrontendService.scala:190)
        at org.apache.kyuubi.service.ThriftFrontendService.OpenSession(ThriftFrontendService.scala:199)
        at org.apache.hive.service.rpc.thrift.TCLIService$Processor$OpenSession.getResult(TCLIService.java:1377)
        at org.apache.hive.service.rpc.thrift.TCLIService$Processor$OpenSession.getResult(TCLIService.java:1362)
        at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
        at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
        at org.apache.kyuubi.service.authentication.TSetIpAddressProcessor.process(TSetIpAddressProcessor.scala:36)
        at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:286)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused (Connection refused)
        at org.apache.thrift.transport.TSocket.open(TSocket.java:226)
        at org.apache.thrift.transport.TSaslTransport.open(TSaslTransport.java:266)
        at org.apache.thrift.transport.TSaslClientTransport.open(TSaslClientTransport.java:37)
        at org.apache.kyuubi.session.KyuubiSessionImpl.openSession(KyuubiSessionImpl.scala:74)
        at org.apache.kyuubi.session.KyuubiSessionImpl.$anonfun$open$2(KyuubiSessionImpl.scala:64)
        at org.apache.kyuubi.session.KyuubiSessionImpl.$anonfun$open$2$adapted(KyuubiSessionImpl.scala:62)
        at org.apache.kyuubi.ha.client.ZooKeeperClientProvider$.withZkClient(ZooKeeperClientProvider.scala:74)
        at org.apache.kyuubi.session.KyuubiSessionImpl.open(KyuubiSessionImpl.scala:62)
        at org.apache.kyuubi.session.KyuubiSessionManager.openSession(KyuubiSessionManager.scala:58)
        ... 12 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:607)
        at org.apache.thrift.transport.TSocket.open(TSocket.java:221)
        ... 20 more
```

Could you give me some tips about this error?
Thanks

**Reply by pan3793** [reply]
OK, I think I know what happens.

Looks like that your Spark cluster running in standalone mode, then when you run `spark-shell` or `spark-sql`, the driver runs in the current machine, but when run `spark-submit` to deploy a jar job, it will random(or based on some algorithm) pick up one machine from your standalone cluster to start the driver process.

Basically, Kyuubi plays a gateway or proxy role of Spark SQL, when you use beeline to connect Kyuubi, it will pick or create an engine(which actually a Spark Jar App, created by `spark-submit`, you can find the full command from your Kyuubi server logs).

Zookeeper plays a discovery service role in this scenario. When the engine started, expose a Thrift endpoint and register itself into Zookeeper, Kyuubi server can get the connection info from Zookeeper and establish the connection to the engine.

We recommend using an external Zookeeper cluster for production, and will start an embedded Zookeeper instance if the Zookeeper confs is absent. If my analysis is correct, there are 2 options that can solve your problem

1. (recommended for production) Modify `kyuubi.ha.zookeeper.quorum` to use external Zookeeper cluster.
   Ref doc: https://kyuubi.apache.org/docs/latest/deployment/settings.html#ha
2. Modify `kyuubi.zookeeper.embedded.client.port.address` to change embedded Zookeeper listen to a host/ip which can be accessed by whole nodes of your spark cluster. 
   Ref doc: https://kyuubi.apache.org/docs/latest/deployment/settings.html#zookeeper

**Reply by zhichaoleo** [reply]
Hi @pan3793，
Thanks for your tips. Actually, I use our zookeeper in my local, and already set properties in kyuubi-defaults.xml like below:
`
spark.master                    spark://******:7077
kyuubi.authentication           NONE
kyuubi.frontend.bind.host       ******
kyuubi.frontend.bind.port       10009
kyuubi.ha.zookeeper.quorum      ******:2181
`

I got the engine service host and port in my zookeeper and test with telnet. And I found the error may caused by DNS mapping error, So I check my /etc/hosts file and found some IPv6 configs which may cause mapping error. After I comment the IPv6 releated properties in /ec/hosts. Beeline can connect to Kyuubi. 

Thanks for your kindly help!

**Reply by yaooqinn** [reply]
Hi @zhichaoleo thank you try Apache Kyuubi and the discussion here looks amazing! 

As you mentioned that

> But I can't find any documents to set up related configurations on kyuubi offical doc site or google search.

So, can you help us build a doc for integrating with delta lake?

**Reply by zhichaoleo** [reply]
Hi @yaooqinn,
Sure. I think I can write down the steps. But I don't know how to make this contribution. Could you show me the steps to build this doc?
Thanks

**Reply by yaooqinn** [reply]
You can fork this repo and clone it to your local machine, then you can open it in any IDE or editor software you like.

The newly add doc should be written in markdown form and `docs/integrations` might be a good place to put it.

The content of your article can be how to integrate Kyuubi and delta lake step by step.

There is also an example for kudu integration in https://github.com/apache/incubator-kyuubi/tree/master/docs/integrations

FYi, it's also easy to verify how the doc looks like in local by https://kyuubi.apache.org/docs/latest/develop_tools/build_document.html

**Reply by zhichaoleo** [reply]
Great. I will write this doc and make a PR.  Hope you and team members can give feedback and suggestions.
Kyuubi is a great tool, and I wish it will get better and better.

**Reply by yaooqinn** [reply]
thanks @zhichaoleo
